ArduPlane/fence.cpp

```diff
@@ -0,0 +1,132 @@
+#include "Plane.h"
+
+// Code to integrate AC_Fence library with main ArduPlane code
+
+#if AC_FENCE == ENABLED
+
+// fence_check - ask fence library to check for breaches and initiate the response
+// called at 1Hz
+void Plane::fence_check()
+{
+    const uint8_t orig_breaches = fence.get_breaches();
+
+    // check for new breaches; new_breaches is bitmask of fence types breached
+    const uint8_t new_breaches = fence.check();
+
+    // we still don't do anything when disarmed, but we do check for fence breaches.
+    // fence pre-arm check actually checks if any fence has been breached
+    // that's not ever going to be true if we don't call check on AP_Fence while disarmed
+    if (!arming.is_armed())
+    {
+        return;
+    }
+
+    if (!fence.enabled())
+    {
+        // Switch back to the chosen control mode if still in
+        // GUIDED to the return point
+        if ((fence.get_action() == FENCE_ACTION_GUIDED || fence.get_action() == FENCE_ACTION_GUIDED_THR_PASS || fence.get_action() == FENCE_ACTION_RTL) &&
```

Comment:
I wonder if this should go before the `is_armed()` check...

Also.... this (preexisting) line sucks.  I wonder if we should just look at `plane.control_mode_reason` and switch back if that was `FenceBreach`

A switch on get_action might also help.

Reply:
Actually points at a bigger issue - that uses `FENCE_ACTION` as what's defined in `mavlink`, and not the `AC_Fence` versions `AC_FENCE_ACTION`. I'll update that accordingly.

The way I read the old implementation is "if we are in a guided or avoid adsb mode because of a fence breach, and the fence is switched off, return to the previous mode."